Update types and props for compatibility with React 18

ggdouglas opened the pull request:
Copied from **React 18 Upgrade feature branch** https://github.com/palantir/blueprint/pull/7142

#### Proposed changes:

Updates types for forward compatibility with React 18.

ggdouglas committed:
Fix compile break in SVGIconContainer

ggdouglas committed:
Fix compile break on React.createElement in ContextMenu

ggdouglas committed:
Fix compile break in EntityTitle

ggdouglas committed:
Fix compile break in HotkeysProps

ggdouglas committed:
Fix compile break on Icon

ggdouglas committed:
Fix compile break in Overlay

ggdouglas committed:
Fix compile break from missing children in ColumnHeaderProps

ggdouglas committed:
Fix compile break caused by ResizeHandle in HeaderCell

ggdouglas committed:
Use inferred return type for DatePicker3Caption

Fixes type error in NonContiguousDayRangePicker related to react-day-picker Caption component prop:
```
Type 'FC<CaptionProps>' is not assignable to type '(props: CaptionProps) => Element | null'.
  Type 'ReactNode' is not assignable to type 'Element | null'.
    Type 'undefined' is not assignable to type 'Element | null'.
```

ggdouglas committed:
Add missing placeholder prop to HTMLSelect props

ggdouglas committed:
Specify type in SegmentedControlExample to avoid implicit any

ggdouglas committed:
Revert "Fix compile break in SVGIconContainer"

This reverts commit 77bf2861ae953312e06f4a6870aa8d99bd2a8196.

ggdouglas committed:
Revert "Fix compile break on Icon"

This reverts commit 67a68f9c6786a0c8b6e6f491b1ded798910e7979.

gluxon commented on packages/core/src/components/entity-title/entityTitle.tsx:
```diff
@@ -177,7 +177,7 @@ EntityTitle.displayName = `${DISPLAYNAME_PREFIX}.EntityTitle`;
  * Construct header class name from H{*}. Returns `undefined` if `heading` is not a Blueprint heading.
  */
-function getClassNameFromHeading(heading: React.FC<unknown>) {
+function getClassNameFromHeading(heading: React.FC<any>) {
```
It can be valuable to use `unknown` rather than `any` since the former is more future-proof. If someone refactors this function, they might not realize there's some type safety loss with the argument.

Does `() => React.ReactNode` work as an alternative to `React.FC<unknown>`?

ggdouglas replied:
Ack, will try this out!

ggdouglas replied:
Changing to `() => React.ReactNode` does not seem to work:
```
Argument of type 'FC<any> | FC<TextProps>' is not assignable to parameter of type '() => ReactNode'.
  Type 'FC<any>' is not assignable to type '() => ReactNode'.
    Target signature provides too few arguments. Expected 1 or more, but got 0.ts(2345)
```
but I do think this leads us to a better solution to use `FC<TextProps>` instead. This function is only used internally in EntityTitle.
```suggestion
function getClassNameFromHeading(heading: React.FC<TextProps>) {
```

ggdouglas replied:
f50354994ce29779ce51af89b0d30f87d4b527de

gluxon replied:
Nice!

gluxon commented on packages/core/src/components/hotkeys/hotkeys.tsx:
```diff
@@ -24,7 +24,7 @@ import { H4 } from "../html/html";
 import { Hotkey, type HotkeyProps } from "./hotkey";
 
-export interface HotkeysProps extends Props {
+export interface HotkeysProps extends Props, React.PropsWithChildren<object> {
```
It looks like we're expecting children to be `React.ReactElement<HotkeyProps>`. Should we use that type over `object`?

https://github.com/palantir/blueprint/blob/0c815e909e86f511d9b057e7fa4f2dd80c91a3c8/packages/core/src/components/hotkeys/hotkeys.tsx#L57

Or was the goal to preserve the same type definition of `HotkeyProps` and avoid compile breaks on consumers?

gluxon replied:
This one is a nitpick, but I've personally been avoiding `React.PropsWithChildren<...>` since I see the React docs usually write the `children` prop directly.

https://react.dev/learn/typescript#typing-children

Having the `children` prop directly in the interface also makes it easier to add a JSDoc on the field.

ggdouglas replied:
fd385b9a1e5e30308df5d7637f2f0aa279ae493b

gluxon commented on packages/table/src/headers/columnHeader.tsx:
```diff
@@ -36,7 +36,7 @@ export interface ColumnWidths {
 }
 
-export interface ColumnHeaderProps extends HeaderProps, ColumnWidths, ColumnIndices {
+export interface ColumnHeaderProps extends React.PropsWithChildren<HeaderProps>, ColumnWidths, ColumnIndices {
```
nitpick: Similar comment around potentially writing the `children` field directly.

ggdouglas replied:
21a62fcc8a12121739b04430f25dc05d1e14da19

ggdouglas committed:
Specify argument type on getClassNameFromHeading

ggdouglas committed:
Prefer direct children type declaration in HotkeysProps

ggdouglas committed:
Prefer direct children type declaration in HeaderProps

ggdouglas committed:
Switch to React.ReactNode for HotkeysProps children declaration

Fixing type error:
```
../table/src/cell/editableCell.tsx:198:17 - error TS2322: Type 'Element' is not assignable to type '(ReactElement<HotkeyProps, string | JSXElementConstructor<any>>[] & (boolean | ReactChild | ReactFragment | ReactPortal | null)) | undefined'.
  Type 'ReactElement<any, any>' is not assignable to type 'ReactElement<HotkeyProps, string | JSXElementConstructor<any>>[] & ReactPortal'.
    Type 'ReactElement<any, any>' is missing the following properties from type 'ReactElement<HotkeyProps, string | JSXElementConstructor<any>>[]': length, pop, push, concat, and 29 more.

198                 <Hotkey
                    ~~~~~~~
199                     key="edit-cell"
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
203                     onKeyDown={this.handleEdit}
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
204                 />
    ~~~~~~~~~~~~~~~~~~

  ../core/lib/esm/components/hotkeys/hotkeys.d.ts:17:5
    17     children?: Array<React.ReactElement<HotkeyProps>>;
           ~~~~~~~~
    The expected type comes from property 'children' which is declared here on type 'IntrinsicAttributes & IntrinsicClassAttributes<Hotkeys> & Pick<Readonly<HotkeysProps> & Readonly<...>, "className" | "children"> & InexactPartial<...> & InexactPartial<...>'

../table/src/table.tsx:504:25 - error TS2322: Type '(Element | Element[] | undefined)[]' is not assignable to type '(ReactElement<HotkeyProps, string | JSXElementConstructor<any>>[] & (boolean | ReactChild | ReactFragment | ReactPortal | null)) | undefined'.
  Type '(Element | Element[] | undefined)[]' is not assignable to type 'ReactElement<HotkeyProps, string | JSXElementConstructor<any>>[]'.
    Type 'Element | Element[] | undefined' is not assignable to type 'ReactElement<HotkeyProps, string | JSXElementConstructor<any>>'.
      Type 'undefined' is not assignable to type 'ReactElement<HotkeyProps, string | JSXElementConstructor<any>>'.

504         return <Hotkeys>{hotkeys.filter(element => element !== undefined)}</Hotkeys>;
                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  ../core/lib/esm/components/hotkeys/hotkeys.d.ts:17:5
    17     children?: Array<React.ReactElement<HotkeyProps>>;
           ~~~~~~~~
    The expected type comes from property 'children' which is declared here on type 'IntrinsicAttributes & IntrinsicClassAttributes<Hotkeys> & Pick<Readonly<HotkeysProps> & Readonly<...>, "className" | "children"> & InexactPartial<...> & InexactPartial<...>'
```

svc-palantir-github commented:
<h3>Switch to React.ReactNode for HotkeysProps children declaration</h3>
Build artifact links for this commit: <strong><a href="https://output.circle-artifacts.com/output/job/42d7b2f4-5e5b-47d0-a434-4b13f7e6eafd/artifacts/0/packages/docs-app/dist/index.html">documentation</a> | <a href="https://output.circle-artifacts.com/output/job/42d7b2f4-5e5b-47d0-a434-4b13f7e6eafd/artifacts/0/packages/landing-app/dist/index.html">landing</a> | <a href="https://output.circle-artifacts.com/output/job/42d7b2f4-5e5b-47d0-a434-4b13f7e6eafd/artifacts/0/packages/table-dev-app/dist/index.html">table</a> | <a href="https://output.circle-artifacts.com/output/job/42d7b2f4-5e5b-47d0-a434-4b13f7e6eafd/artifacts/0/packages/demo-app/dist/index.html">demo</a></strong>

<em>This is an automated comment from the deploy-preview CircleCI job.</em>
